### PR TITLE
Upgrade commitlint to 17.4.0 version without upgrading typescript to 4.9.4

### PR DESCRIPTION
**Pull request by realmarv**

### CODE REVIEW

**Comment by knocte** [comment]
@realmarv nit: please replace all single-dash flags with double-dash ones, for example --yes instead of -y, or --global instead of -g (is it global? I guess)

**Comment by knocte** [comment]
2nd thing: why didn't you push 1 by 1? If the answer to this is: "because some commits might fail CI", then to that I would answer: "then it sounds like you should squash more, if not all?".

**Comment by knocte** [comment]
3rd: in the commit msg that updates the package-* files, please state clearly in the body of the message the `npm install` invocations that you performed in order to arrive to that state.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -14,15 +14,17 @@ jobs:
...
+          curl -sL https://deb.nodesource.com/setup_18.x | sudo -E bash -
+          sudo DEBIAN_FRONTEND=noninteractive apt install -y nodejs
+
+      - name: Install conventions dependencies
         run: |
-          git --version
-          node --version
-          npm --version
-          npx commitlint --version
+          npm install -g typescript@4.8.4
```
Comment (knocte) [inline .github/workflows/ci.yml:26]: @realmarv what happens if we use 4.9.4 instead of 4.8.4?

**Comment by realmarv** [comment]
> 2nd thing: why didn't you push 1 by 1? If the answer to this is: "because some commits might fail CI", then to that I would answer: "then it sounds like you should squash more, if not all?".

I pushed 1 by 1. I don't know why github only shows the CI result for the last commit but if you click on the previous commits you can see the status of CI there.

File: package.json [hunk 2]
```diff
@@ -1,26 +1,321 @@
...
+    "yallist": "^4.0.0",
+    "yargs": "^17.6.0",
+    "v8-compile-cache-lib": "^3.0.1",
+    "validate-npm-package-license": "^3.0.4",
+    "yargs-parser": "^21.1.1",
+    "write-file-atomic": "^4.0.2",
+    "yocto-queue": "^0.1.0",
+    "walker": "^1.0.8",
+    "wrappy": "^1.0.2",
+    "@commitlint/config-conventional": "^17.1.0"
```
Comment (knocte) [inline package.json:296]: @realmarv this doesn't look correct, the title of this PR states version 17.4.0, not 17.1.0

**Comment by knocte** [comment]
> I pushed 1 by 1. I don't know why github only shows the CI result for the last commit but if you click on the previous commits you can see the status of CI there.

Arghhh, damned github! Anyway, I took a look at the `Actions` tab of your fork and I saw that the previous to last commit has CI red. Which means that my point still stands.

File: package.json (hunk 2 above)
Comment (realmarv) [reply package.json:296]: This is one of the commitlint dependencies.  Commitlint's version is 17.4.0. But I can upgrade this one as well. I hope it doesn't break the CI

Comment (knocte) [reply package.json:296]: >  I hope it doesn't break the CI

Try it in different branch first?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (realmarv) [reply .github/workflows/ci.yml]: It broke the CI

Comment (knocte) [reply .github/workflows/ci.yml]: then you didn't really fix the typescript problem, you managed to upgrade commitlint without upgrading typescript :) (which is kinda ok, for now, just rename the PR to make it clear)

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -53,18 +67,31 @@ jobs:
...
+          tsc commitlintplugins.test.ts --target es2017 --moduleResolution node
       - name: Validate current commit (last commit) with commitlint
         if: github.event_name == 'push'
         run: ./commitlint.sh --from HEAD~1 --to HEAD --verbose
       - name: Validate PR commits with commitlint
         if: github.event_name == 'pull_request'
         run: ./commitlint.sh --from ${{ github.event.pull_request.head.sha }}~${{ github.event.pull_request.commits }} --to ${{ github.event.pull_request.head.sha }} --verbose
       - name: Install dotnet sdk
-        run: sudo apt install -y dotnet6
+        run: sudo apt install --yes dotnet6
```
[the remark this answers is not in the document]
Comment (realmarv) [reply .github/workflows/ci.yml]: Done

**Comment by realmarv** [comment]
> @realmarv nit: please replace all single-dash flags with double-dash ones, for example --yes instead of -y, or --global instead of -g (is it global? I guess)

Done

**Comment by realmarv** [comment]
> 3rd: in the commit msg that updates the package-* files, please state clearly in the body of the message the `npm install` invocations that you performed in order to arrive to that state.

Done

**Comment by knocte** [comment]
> is generated by `npm i --package-lock-only` commands locally.

What commands? Please list them.

**Comment by realmarv** [comment]
> > I pushed 1 by 1. I don't know why github only shows the CI result for the last commit but if you click on the previous commits you can see the status of CI there.
> 
> Arghhh, damned github! Anyway, I took a look at the `Actions` tab of your fork and I saw that the previous to last commit has CI red. Which means that my point still stands.

I squashed those import fixes. Do I squash more?

**Comment by knocte** [comment]
> I squashed those import fixes. Do I squash more?

The existence of "Revert commits" tell me that the answer to this is YES :)

**Comment by realmarv** [comment]
> > I squashed those import fixes. Do I squash more?
> 
> The existence of "Revert commits" tells me that the answer to this is YES :)

Done

File: package.json (hunk 2 above)
Comment (knocte) [reply package.json:296]: this is still not addressed ^ pls do next day

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -14,15 +14,17 @@ jobs:
...
           sudo apt install -y git
-          sudo DEBIAN_FRONTEND=noninteractive apt install --yes npm
-      - name: Install commitlint
-        run: npm install
-      - name: Print versions
+          
+          sudo apt install --yes curl
+          # can't install ubuntu's default nodejs version because we would get this error:
+          # error @commitlint/config-conventional@17.1.0: The engine "node" is incompatible with this module. Expected version ">=v14". Got "12.22.9"
+          curl -sL https://deb.nodesource.com/setup_18.x | sudo -E bash -
```
Comment (knocte) [inline .github/workflows/ci.yml:21]: @realmarv by the error message above it sounds like we don't need to be in such a bleeding edge version, let's try 14.x instead

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -32,8 +34,14 @@ jobs:
           node --version
           npm --version
           npx commitlint --version
+      - name: Compile ts files
```
Comment (knocte) [inline .github/workflows/ci.yml:37]: @realmarv ts is not a blessed abbreviation in our list :) which means you probably meant to use `TS` (acronym) or `typescript` or `.ts`

Comment (realmarv) [reply .github/workflows/ci.yml:37]: Done

File: .github/workflows/ci.yml (hunk 4 above)
Comment (realmarv) [reply .github/workflows/ci.yml]: It broke the CI https://github.com/realmarv/conventions/actions/runs/3928561523/jobs/6716315668

File: package.json (hunk 2 above)
Comment (realmarv) [reply package.json:296]: Tried here: https://github.com/realmarv/conventions/commits/config-conventional-17.4.0
It seems ok

File: .github/workflows/ci.yml (hunk 4 above)
Comment (knocte) [reply .github/workflows/ci.yml]: then 16.x?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (realmarv) [reply .github/workflows/ci.yml]: Not sure. You said it's ok for now not totally, so I thought we should upgrade the typescript as well

File: .github/workflows/ci.yml (hunk 4 above)
Comment (realmarv) [reply .github/workflows/ci.yml]: Done

Comment (realmarv) [reply .github/workflows/ci.yml]: But why is the reason?

File: package.json (hunk 2 above)
Comment (realmarv) [reply package.json:296]: Do I apply the changes here?

File: .github/workflows/ci.yml (hunk 4 above)
Comment (knocte) [reply .github/workflows/ci.yml]: I already said it above; do you understand what "bleeding edge" means in terms of versions? it means too high, potentially unstable; lower versions have been tested more, so more stable

**Comment by knocte** [comment]
> I mentioned the two commands that I used. I'm not sure what you mean 🤔

I'm sure those commands are not enough, you did something else. To prove my point, I tested it myself: I cloned conventions repo in my Linux VM, I typed those 2 commands, and they didn't cause any modifications to the `dependencies` part of the `package.json` file.

**Comment by realmarv** [comment]
> > I mentioned the two commands that I used. I'm not sure what you mean thinking
> 
> I'm sure those commands are not enough, you did something else. To prove my point, I tested it myself: I cloned conventions repo in my Linux VM, I typed those 2 commands, and they didn't cause any modifications to the `dependencies` part of the `package.json` file.

I think package.json is generated based on the modules that are already installed (They are in the node_modules folder.), so when you clone the repository, there are no node_modules until you install them using `npm install` based on the dependencies defined in package.json. I had those packages installed on my computer so I generated package.json and package-lock.json based on them.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (realmarv) [reply .github/workflows/ci.yml]: Got it thanks

**Comment by knocte** [comment]
> I think package.json is generated based on the modules that are already installed (They are in the node_modules folder.), so when you clone the repository, there are no node_modules until you install them using npm install based on the dependencies defined in package.json

Not sure if what you are saying is 100% accurate, but let's suppose it is for now. Then:

> I had those packages installed on my computer

So, what commands did you use to get those packages installed in your computer?

**Comment by realmarv** [comment]
> > I think package.json is generated based on the modules that are already installed (They are in the node_modules folder.), so when you clone the repository, there are no node_modules until you install them using npm install based on the dependencies defined in package.json
> 
> Not sure if what you are saying is 100% accurate, but let's suppose it is for now. Then:
> 
> > I had those packages installed on my computer
> 
> So, what commands did you use to get those packages installed in your computer?

I had installed them when I started working on this repo, so I don't really remember what was the exact procedure. I just upgraded commitlint last week :thinking:

**Comment by knocte** [comment]
This PR is superseded by #55 right?

**Comment by realmarv** [comment]
This PR is closed and replaced by https://github.com/nblockchain/conventions/pull/55.

**Comment by knocte** [comment]
> This PR is replaced by ...

I prefer the word "superseded" :)

**Comment by realmarv** [comment]
> > This PR is replaced by ...
> 
> I prefer the word "superseded" :)

Ok

**Pull request closed without merging**